### ISSUE TITLE
Callback on scrape initiation

### ISSUE THREAD

**Issue by hhoffstaette** [jupp0r/prometheus-cpp#224]
One thing I'm wondering about is how to trigger counter updates when there is no other
activity in the "sleep loop" (as per example). In my case I only need to update the metrics
on-demand, but right now there doesn't seem to be a way to be notified of this.
I *could* run a second control loop with timed waiting (passing the update delay into main()),
but that would expose stale data when the scrape and the update loop diverge, and they
inevitably will..

**Comment by jupp0r** [comment]
You could cache the latest result of `Collect()`, compare it with the newest result and only push if they differ. At this point, `prometheus-cpp` does not provide a lot of support for this use case and only offers fairly low-level tools. If there is a common pattern that people implement on top, I'm more than willing to add this to the library, but at the same time I'd prefer the library to not have an opinion on things like threading and timers as much as possible.

**Comment by gjasny** [comment]
AFAIU you'd like to update counters lazily on collect? Maybe implementing your own collectable would help here. As we see in the scalable counter PR there are different needs for different strategies.

**Comment by hhoffstaette** [comment]
> AFAIU you'd like to update counters lazily on collect?

Yes, exactly. The endless for() loop with sleep() as outlined in the README is
essentially "always wrong" - how long am I supposed to sleep when updating counters
is expensive (compared to the rest of the lib)?

> Maybe implementing your own collectable would help here.

I was just poking through the source and was wondering somethign similar, though
I'd opt for simply being able to pass in a lambda to either the exposer or the registry to call before
their Collect()s.

**Comment by jupp0r** [comment]
I'm not sure what your goal is @hhoffstaette . The `sleep` inside README.md is only meant as a placeholder for some actual work your server application would perform, not something you would need to put in any of your code. What do you want to measure?

If you wanted to be informed of scrapes, you could implement something along the lines of

```
class CollectableDelegate : public prometheus::Collectable {
public:
    CollectableDelegate(std::unique_ptr<Collectable> collectable) : collectable_{std::move(collectable)} {}
   auto Collect() override {
        // your custom code here
        return collectable_->Collect();
   }
```

construct a `CollectableDelegate` by passing a `Registry` in and then register the `CollectableDelegate` with an `Exposer` as you'd normally register the `Registry`.

**Comment by hhoffstaette** [comment]
> I'm not sure what your goal is @hhoffstaette . The `sleep` inside README.md is only meant as a
> placeholder for some actual work your server application would perform

Yes, I understand that. My point is that this placeholder is indicative of a missing piece
of the puzzle (the callback I am after). We have two control loops - the web scrape and
whatever is running in the example for() loop. But those two loops are disconnected
in time.

> If you wanted to be informed of scrapes, you could implement something along the lines of

Well yes, thanks! That is exactly what I was looking for! :)
More precisely I was just wondering how it might fit in with the existing design - adding a
callback somewhere vs. overriding/extending Registry vs. delegating to it etc.

I'll give the wrapper thing a try and see how it works (might be a few days).
Something like this would make for a good example, or could even be part of the lib.

**Comment by hhoffstaette** [comment]
So this approach with implementing Collect() on my own counter collections worked well. I now have a delegate which manages a set of counters (one for each found mountpoint) and an aggregate delegate which manages them. The aggregate is registered with the Exposer, and on Collect() tickles all existing (or new) sub-delegates. They update themselves, return the counter sets and the aggregate returns them all.
This has also accidentally fixed the problem I raised in #178 since when a mountpoint appears/disappears I can just create/destroy an entire counter set instead of mutating a Registry.

**Comment by jupp0r** [comment]
Nice, just keep in mind you'd still leak the memory in the registry, though.
